### ISSUE TITLE
Possible Memory Leak While Inferencing for an NER Task

### ISSUE THREAD

**Issue by morgworth** [NVIDIA/NeMo#631]
I am currently working on a project that incorporates NeMo for an NER task and transfer learns the weights for the default bert_base_uncased model. The training is complete and we are able to inference with the trained model. The max sequence length of training inputs and inferencing queries is set to 128. We are using a single Tesla V100 GPU and are using batch inferencing with a batch size of 4096. 

We are looking to inference on ~76.5 million queries. I first tried to send all the queries at once but ran into an issue of all of the memory being used up on the DGX 1 machine we are using for development. 

Then, I worked on modifying the script /NeMo/examples/nlp/token_classification/token_classification_infer.py so that it can be called as a function. I then use this function to inference on groupings of 100,000 queries and writing those to a text file in another script. This script, inference_per_100000.py, still stops running eventually and the memory usage is high. It may be a memory leak in NeMo itself.

I have attached the two scripts described above and a screenshot of the memory usage from the tool htop.

Maybe there are better ways to handle:
1. writing the inferencing output
2. passing in the queries in for inferencing

If that is the case, and it is not a memory leak,  I can work on implementing those improvements in my scripts. 
[inference_per_100000.txt](https://github.com/NVIDIA/NeMo/files/4624530/inference_per_100000.txt)
[token_classification_infer.txt](https://github.com/NVIDIA/NeMo/files/4624531/token_classification_infer.txt)
![Potential NeMo Memory Leak](https://user-images.githubusercontent.com/33635540/81862052-faa61500-951d-11ea-8deb-cd83aba8ed58.jpg)

**Comment by yzhang123** [comment]
@morgworth thanks for letting us know about this issue. 

I can see from your changed token_classification_infer.txt that you are 
creating a new neural graph in inference() which means every 100k steps, this will increase memory. Please create the modules AND neural graph once outside the function and let me know if it works?

**Comment by yzhang123** [comment]
@morgworth after looking at your issue more closely i was able to reproduce your growing memory problem. The issue is with every loop (call of inference) you are increasing the objects in your default neural graph without deleting the previous graph. So one quick solution would be to declare an explicit graph and delete that together with neural modules you define in every new iteration (e.g. data_layer)
1. Wrap your graph into an explicit graph:

    ```
     with NeuralGraph(operation_mode=OperationMode.evaluation) as g1:
             input_ids, input_type_ids, input_mask, _, subtokens_mask = data_layer()
             hidden_states = model(input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask)
             logits = classifier(hidden_states=hidden_states)```

2. move `classifier = TokenClassifier(hidden_size=hidden_size, num_classes=len(labels_dict))` out of the inference function
3. delete graph and modules defined in each loop
```
    del data_layer
    del g1
```

**Comment by tkornuta-nvidia** [comment]
@morgworth For more information on Neural Graphs you can try going through the Jupyter Notebook tutorials:
https://github.com/NVIDIA/NeMo/tree/master/examples/neural_graphs
You can ping me directly if there will be anything unclear.

**Comment by morgworth** [comment]
@yzhang123 @tkornuta-nvidia Thanks for your assistance in looking into this. I have implemented the change and will test it soon. I will follow up with the results from testing.

**Comment by wjyfelicity** [comment]
@tkornuta-nvidia , I am currently working on a project that incorporates NeMo for an asr task, when infer the result, there is a memory leak．

1. infer with language model

![image](https://user-images.githubusercontent.com/24841439/84462705-488b7580-aca2-11ea-9733-b8dd2d3dec6f.png)　

2. infer without language model
![image](https://user-images.githubusercontent.com/24841439/84462783-7b356e00-aca2-11ea-806c-0610e69a55e1.png)

and the inference code is :
class ModelASR():
    def __init__(self, args):
        parser = get_parser()
        args = parser.parse_args(args)
        self.load_model(args)
    
    @profile
    def load_model(self, args):
        batch_size = args.batch_size
        load_dir = args.load_dir

        if args.local_rank is not None:
            if args.lm_path:
                raise NotImplementedError(
                    "Beam search decoder with LM does not currently support evaluation on multi-gpu."
                )
            device = nemo.core.DeviceType.AllGpu
        else:
            device = nemo.core.DeviceType.GPU

        # Instantiate Neural Factory with supported backend
        neural_factory = nemo.core.NeuralModuleFactory(
            backend=nemo.core.Backend.PyTorch,
            local_rank=args.local_rank,
            placement=device,
        )

        if args.local_rank is not None:
            logging.info('Doing ALL GPU')

        yaml = YAML(typ="safe")
        with open(args.model_config) as f:
            jasper_params = yaml.load(f)
        vocab = jasper_params['labels']
        sample_rate = jasper_params['sample_rate']

        data_preprocessor = nemo_asr.AudioToMelSpectrogramPreprocessor(
            sample_rate=sample_rate, **jasper_params["AudioToMelSpectrogramPreprocessor"]
        )
        jasper_encoder = nemo_asr.JasperEncoder(
            feat_in=jasper_params["AudioToMelSpectrogramPreprocessor"]["features"], **jasper_params["JasperEncoder"]
        )
        jasper_decoder = nemo_asr.JasperDecoderForCTC(
            feat_in=jasper_params["JasperEncoder"]["jasper"][-1]["filters"], num_classes=len(vocab)
        )
        greedy_decoder = nemo_asr.GreedyCTCDecoder()

        logging.info('================================')
        logging.info(f"Number of parameters in encoder: {jasper_encoder.num_weights}")
        logging.info(f"Number of parameters in decoder: {jasper_decoder.num_weights}")
        logging.info(f"Total number of parameters in model: " f"{jasper_decoder.num_weights + jasper_encoder.num_weights}")
        logging.info('================================')


        if args.lm_path:
            if args.alpha_max is None:
                args.alpha_max = args.alpha
            # include alpha_max in tuning range
            args.alpha_max += args.alpha_step / 10.0

            if args.beta_max is None:
                args.beta_max = args.beta
            # include beta_max in tuning range
            args.beta_max += args.beta_step / 10.0

        # language model
            for alpha in np.arange(args.alpha, args.alpha_max, args.alpha_step):
                for beta in np.arange(args.beta, args.beta_max, args.beta_step):
                    logging.info('================================')
                    logging.info(f'Infering with (alpha, beta): ({alpha}, {beta})')
                    beam_search_with_lm = nemo_asr.BeamSearchDecoderWithLM(
                        vocab=vocab,
                        beam_width=args.beam_width,
                        alpha=alpha,
                        beta=beta,
                        lm_path=args.lm_path,
                        num_cpus=max(os.cpu_count(), 1),
                        input_tensor=False,
                    )
            self.beam_search_with_lm = beam_search_with_lm
                    

        self.data_preprocessor = data_preprocessor
        self.jasper_encoder = jasper_encoder
        self.jasper_decoder = jasper_decoder
        self.greedy_decoder = greedy_decoder
        self.neural_factory = neural_factory
        self.vocab = vocab
        self.args = args
        self.sample_rate = sample_rate
        
    def infer(self, wav_path, json_file):
        get_feature_json(wav_path, json_file)
        eval_datasets=json_file
        sample_rate=self.sample_rate
        vocab=self.vocab
        batch_size=1
        data_layer = nemo_asr.AudioToTextDataLayer(
            manifest_filepath=eval_datasets,
            sample_rate=sample_rate,
            labels=vocab,
            batch_size=batch_size,
        )
        
        audio_signal_e1, a_sig_length_e1, transcript_e1, transcript_len_e1 = data_layer()
        processed_signal_e1, p_length_e1 = self.data_preprocessor(input_signal=audio_signal_e1, length=a_sig_length_e1)
        encoded_e1, encoded_len_e1 = self.jasper_encoder(audio_signal=processed_signal_e1, length=p_length_e1)
        log_probs_e1 = self.jasper_decoder(encoder_output=encoded_e1)
        predictions_e1 = self.greedy_decoder(log_probs=log_probs_e1)
        eval_tensors = [log_probs_e1, predictions_e1, transcript_e1, transcript_len_e1, encoded_len_e1]

        # inference
        evaluated_tensors = self.neural_factory.infer(tensors=eval_tensors, checkpoint_dir=self.args.load_dir)
        
        greedy_hypotheses = post_process_predictions(evaluated_tensors[1], self.vocab)
        
        logprob = []
        for i, batch in enumerate(evaluated_tensors[0]):
            for j in range(batch.shape[0]):
                logprob.append(batch[j][: evaluated_tensors[4][i][j], :].cpu().numpy())

        del evaluated_tensors
        del data_layer
        if self.args.lm_path: 
            logprob = [np.exp(p) for p in logprob]
            beam_predictions = self.beam_search_with_lm(log_probs=logprob, log_probs_length=None, force_pt=True)
            beam_results = [b[0][1] for b in beam_predictions[0]]          
            return beam_results
        else:
            return greedy_hypotheses

**Comment by tkornuta-nvidia** [comment]
Hi @wjyfelicity I have a hard time finding "the memory leak" in the screenshots attached. It seems that the model without LM is slightly smaller from the one with LM - which makes sense to me.

So what is exactly happening?

**Comment by tkornuta-nvidia** [comment]
Also, any particular reason for creating beam_search_with_lm object many times, in this double for loop? Please note that at the end you store the reference to the last object only...

**Comment by yzhang123** [comment]
@wjyfelicity i created a separate issue for this https://github.com/NVIDIA/NeMo/issues/725

**Comment by yzhang123** [comment]
closing due to inactivity

**Comment by Engineering-Course** [comment]
@tkornuta-nvidia 
I find that the memory leak may come from the BeamSearchDecoderWithLM class in [bean_search_decoer.py](https://github.com/NVIDIA/NeMo/blob/master/nemo/collections/asr/beam_search_decoder.py)

`
@profile

    def forward(self, log_probs, log_probs_length):
        probs_list = log_probs
        if self.input_tensor:
            probs = torch.exp(log_probs)
            probs_list = []
            for i, prob in enumerate(probs):
                probs_list.append(prob[: log_probs_length[i], :])
        res = self.beam_search_func(
            probs_list,
            self.vocab,
            beam_size=self.beam_width,
            cutoff_prob=self.cutoff_prob,
            cutoff_top_n=self.cutoff_top_n,
            ext_scoring_func=self.scorer)
        return [res]
`
I profile this function and find that there will be around 0.5M memory increased after calling forward function of BeamSearchDecoderWithLM. More exactly, it comes from the beam_search_func which is imported from ctc_decoders. https://github.com/NVIDIA/NeMo/blob/3366abbda42a41d200e4d437c24e9874779ed2d0/nemo/collections/asr/beam_search_decoder.py#L68

![image](https://user-images.githubusercontent.com/7713394/84970829-0f458080-b14e-11ea-867b-ebea2e456e00.png)

**Comment by tkornuta-nvidia** [comment]
@Engineering-Course thanks for tracking it down, we need to investigate this. 

But this is clearly a different issue, not related in any way with "unaware" adding modules to "default" Neural Graphs many times that was the main issue of the original bug. So can you please open a separate bug with that description?

**Comment by Engineering-Course** [comment]
@tkornuta-nvidia 
Thanks. It is the same problem as that from @wjyfelicity . #725

**Comment by morgworth** [comment]
Hello,

Sorry for the lapse in response. The issue was resolved. Thanks for the help in suggesting explicitly declaring the graphs.

Thanks again!
